source/common/upstream/od_cds_api_impl.cc

```diff
@@ -0,0 +1,101 @@
+
+#include "common/common/assert.h"
+#include "common/grpc/common.h"
+
+#include "absl/strings/str_join.h"
+
+namespace Envoy {
+namespace Upstream {
+
+OdCdsApiSharedPtr
+OdCdsApiImpl::create(const envoy::config::core::v3::ConfigSource& odcds_config,
+                     OptRef<xds::core::v3::ResourceLocator> odcds_resources_locator,
+                     ClusterManager& cm, Stats::Scope& scope,
+                     ProtobufMessage::ValidationVisitor& validation_visitor) {
+  return OdCdsApiSharedPtr(
+      new OdCdsApiImpl(odcds_config, odcds_resources_locator, cm, scope, validation_visitor));
+}
+
+OdCdsApiImpl::OdCdsApiImpl(const envoy::config::core::v3::ConfigSource& odcds_config,
+                           OptRef<xds::core::v3::ResourceLocator> odcds_resources_locator,
+                           ClusterManager& cm, Stats::Scope& scope,
+                           ProtobufMessage::ValidationVisitor& validation_visitor)
+    : Envoy::Config::SubscriptionBase<envoy::config::cluster::v3::Cluster>(
+          odcds_config.resource_api_version(), validation_visitor, "name"),
+      helper_(cm, "odcds"), cm_(cm), scope_(scope.createScope("odcds.")),
+      status_(StartStatus::NotStarted) {
+  // TODO(krnowak): Move the subscription setup to CdsApiHelper. Maybe make CdsApiHelper a base
+  // class for CDS and ODCDS.
+  const auto resource_name = getResourceName();
+  if (!odcds_resources_locator.has_value()) {
+    subscription_ = cm_.subscriptionFactory().subscriptionFromConfigSource(
+        odcds_config, Grpc::Common::typeUrl(resource_name), *scope_, *this, resource_decoder_, {});
+  } else {
+    subscription_ = cm.subscriptionFactory().collectionSubscriptionFromUrl(
+        *odcds_resources_locator, odcds_config, resource_name, *scope_, *this, resource_decoder_);
+  }
+}
+
+void OdCdsApiImpl::onConfigUpdate(const std::vector<Config::DecodedResourceRef>& resources,
+                                  const std::string& version_info) {
+  UNREFERENCED_PARAMETER(resources);
+  UNREFERENCED_PARAMETER(version_info);
+  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+}
+
+void OdCdsApiImpl::onConfigUpdate(const std::vector<Config::DecodedResourceRef>& added_resources,
+                                  const Protobuf::RepeatedPtrField<std::string>& removed_resources,
+                                  const std::string& system_version_info) {
+  auto exception_msgs =
+      helper_.onConfigUpdate(added_resources, removed_resources, system_version_info);
+  sendAwaiting();
+  if (!exception_msgs.empty()) {
+    throw EnvoyException(
+        fmt::format("Error adding/updating cluster(s) {}", absl::StrJoin(exception_msgs, ", ")));
+  }
+}
+
+void OdCdsApiImpl::onConfigUpdateFailed(Envoy::Config::ConfigUpdateFailureReason reason,
+                                        const EnvoyException*) {
+  ASSERT(Envoy::Config::ConfigUpdateFailureReason::ConnectionFailure != reason);
+  sendAwaiting();
+}
+
+void OdCdsApiImpl::sendAwaiting() {
```

Comment:
I think `sendAwaiting` is confusing: this method advances state *and* clears up update requests accumulated while the subscription was starting. Perhaps worth breaking it up?

Reply:
I'll split out the state advancement to a separate place.